### ISSUE TITLE
feature request: newtype for arbitrary base types

### ISSUE THREAD

**Issue by erniecohen** [dafny-lang/dafny#1201]
I don't know whether there is something that makes this hard, but it would be nice to have newtypes to be based on arbitrary value types, rather than just numeric types. (IOW, allow predicate subtypes for any type.) Except perhaps for type parameters, it seems that the hard parts of doing this are probably done now that newtypes allow member declarations. Today's workaround of a datatype with a single member is bad for various obvious reasons.

**Comment by RustanLeino** [comment]
I agree this would be a nice feature. My hunch is that the trickiest part will be to define type inference, since the types of expressions like `Cons(5, Nil)` and `multiset{2, 5}` couldn't be decided immediately (to allow for their types to be the `newtype` of `List<int>` or `multiset<int>`).

**Comment by MikaelMayer** [comment]
Based on the idea described in [this discussion](https://github.com/dafny-lang/dafny/discussions/1557#discussioncomment-1926952), the following example:

```dafny
newtype OpaqueList = x : List<int> | E(x)
newtype OpaqueMultiset = x : multiset<int> | F(x)

method test() {
  var x := Cons(5, Nil) as OpaqueList;
  var y := multiset{2, 5} as OpaqueMultiset
}
```

would be rewritten in a following version which would be "desugared" into:

```dafny
opaque predicate type OpaqueList(x : List<int>) { E(x) }
opaque predicate type OpaqueMultiset(x : multiset<int>) { F(x) }

method test() {
  var x: OpaqueList = (var l := Cons(5, Nil); reveal OpaqueList(l); assert OpaqueList(l); l);
  var x: OpaqueMultiset = (var m := multiset{2, 5}; reveal OpaqueMultiset(m); assert OpaqueMultiset(m); m);
}
```

**Comment by davidcok** [comment]
+1 for this. I find it would be convenient to have a type of non-empty strings.
